proposals/lip-0053.md

```diff
@@ -252,178 +479,192 @@ crossChainUpdateTransactionParams = {
 
-To execute cross-chain updates, the following is done:
+##### Verification
 
-* If `partnerChain.status == CHAIN_REGISTERED` and `inboxUpdate` is non-empty:
-  * If the first CCM in `inboxUpdate` is a <code>[registration CCM][lip-0049#registrationMessage]</code> `ccm` with `ccm.sendingChainID == patnerChainID`:    
-  set `partnerChain.status = CHAIN_ACTIVE`.
-  * Else: <code>[terminateChainInternal][lip-0045#terminateChainInternal](partnerChainID)</code>. Exit the transaction processing, the CCU has no further effect.
-* For every `ccm` in `inboxUpdate.crossChainMessages`:
-  * If `ownChainAccount.ID != ccm.receivingChainID` or <code>[validateFormat][lip-0049#validateFormat](ccm) != True</code>:  
-  <code>[terminateChainInternal][lip-0045#terminateChainInternal](partnerChainID)</code>.
-  * <code>[appendToInboxTree][lip-0045#appendToInboxTree](ccu.sendingChainID, serializedCCM)</code>, where `serializedCCM` is the serialization of `ccm` using [`crossChainMessageSchema`][lip-0049#ccmSchema].
-  * <code>[apply][lip-0049#apply](ccu, ccm)</code>.
-* Continue the processing by executing all steps in the ["Common Processing" section](#common-processing).
+```python
+def verify(trs: Transaction) -> None:
+    # Only the mainchain can send this command.
+    if trs.params.sendingChainID != MAINCHAIN_ID:
+         raise Exception("Only the mainchain can send a sidechain cross-chain update.")
+    if len(trs.params.certificate) > 0:
+        verifyCertificate(trs)
+    if len(trs.params.activeValidatorsUpdate) > 0 or trs.params.newCertificateThreshold > 0:
+        verifyValidatorsUpdate(trs)
+    if trs.params.inboxUpdate is not empty: # An empty object has all properties set to their default values.
+        verifyChannel(trs)
+```
 
-#### Common Processing
+##### Execution
 
-For CCU transactions posted on the mainchain or on sidechains, once the specific execution steps described above have been completed, the following is done in the given order:
+```python
+def execute(trs: Transaction) -> None:
+    # Verify certificate signature. We do it here because if it fails, the transaction fails rather than being invalid.
+    verifyCertificateSignature(trs)
+    # Process cross-chain messages in inbox update.
+    crossChainMessages = trs.params.inboxUpdate.crossChainMessages
+
+    # First process basic checks for all CCMs.
+    for ccmBytes in crossChainMessages:
+        # Verify general format. Past this point, we can access ccm root properties.
+        try:
+            validateFormat(ccm)
+            # The CCM must come be directed to the sidechain, unless it was bounced on the mainchain.
+            if ownChainAccount.ID != ccm.receivingChainID and ccm.status != CCM_STATUS_CHANNEL_UNAVAILABLE:
+                raise Exception("CCM is not directed to the sidechain.")
+            elif ownChainAccount.ID != ccm.sendingChainID and ccm.status == CCM_STATUS_CHANNEL_UNAVAILABLE:
```

Comment:
If I understand correctly this means that in the case of bouncing, we don't swap sending and receiving chain, but in all other cases we do. Wouldn't this make the error handling for CCMs (those with status `<> CCM_STATUS_OK`) cumbersome as you always need a special case for `CCM_STATUS_CHANNEL_UNAVAILABLE`?

Reply:
We have several comments related to bouncing, maybe we could clarify this points all together and the update the LIP accordingly.

Reply:
Can be resolved as the bouncing is now handled differently.